README.md
Overview
--------

[![Build Status](https://secure.travis-ci.org/walmartlabs/backbone-historytracker.png?branch=master)](http://travis-ci.org/walmartlabs/backbone-historytracker)

History-tracker is a backbone plugin that enhances the history management options available in the
core backbone library.

## History Navigation APIs

This also adds support for

* `Backbone.history.back(trigger)`
* `Backbone.history.forward(trigger)`
* `Backbone.history.go(count, trigger)`

Which mirror their associated `window.history` APIs but provide the ability to trigger or not
trigger the route that the change lands on.

`trigger` may be a truthy value to cause the background route to execute or alternatively may be a
callback of the form `trigger(fragment, route)` that should return truthy to trigger a particular
route after the navigation has occurred.

Note that these calls do not prevent one from leaving the SPA context if the user has other pages
before or after the current page in their history.

## Backbone.history.stepOut

Dealing with iframes within a backbone app is problematic as they push on to your history stack and
under most browsers this is not easily removed from the history by removing the iframe.
`Backbone.history.stepOut(options)` provides a utility to work around this shortcoming by stepping
backward until the iframe's history entries are removed in addition to the current backbone route
entry.

Accepts an `options` hash with the following fields:

* `view`: View element hosting the iframe.
    Any iframes hosted in this view will be removed. It is highly recommended that this is
    passed as it will reduce the chance of side effects from the iframe navigation and
    under [some browsers](https://bugzilla.mozilla.org/show_bug.cgi?id=293417) it will make the
    operation significantly faster.
* `callback(fragment, existingTriggered)`: Callback that is called when the operation is complete
* `trigger`: Same options as the trigger parameter above. Additionally may be a string parameter
    which will ensure the final route is the result of the operation.
* `stepLimit` : Maximum number of steps to take.
    This is designed to prevent runaway operations. Defaults to 10
* `routeLimit`: When used in conjunction with a string `trigger` parameter defines the maximum
    number of steps to try to find the trigger parameter before forcing the new trigger route.
    Defaults to 1 and is independent of the `stepLimit` parameter.

## History Location Tracking

When enabled, this feature will associate an incrementing index with every route the user
visits.  This is beneficial, for example, if you are interested in having a different transition
from routes depending on whether the user has clicked the back button or transitioned to a new
route.

With "push state" support
-------------------------
Every time a new route is visited, the URL state will have the unique index associated

No "push state" support
-----------------------
Every time a new route is visited, the route will be re-written to be prefixed with #{index}#.
The `Backbone.History.getFragment` will return the fragment will return the fragment value without
the additional state.

Usage
-----
To turn on the route indexing, use the `trackDirection` property.

    Backbone.history.start({trackDirection: true});

## Hacks and Workarounds

### Webkit location.replace

https://bugs.webkit.org/show_bug.cgi?id=63777

Webkit has known issues with `location.replace` so when available we opt to use `replaceState`
under Webkit.

### Android Replace Navigation

https://bugs.webkit.org/show_bug.cgi?id=85881

Under Android the replace operation does not properly replace content, leading to corruption of the
history. Net effect is that a series of operations such as:

1. Trigger #foo
2. Replace #bar
3. Trigger #baz

Will corrupt the history producing an actual history of:

1. #foo
2. #baz

This is fixed by making `replace` navigations two step on the devices known to exhibit this behavior.

#### Caveats

This will break sites that attempt to do a replace navigation on the first page. For sites that
absolutely must do this, the `noReplaceHack` flag may be utilized at the expense of broken replace
behavior in these environments.

### Chrome for iOS form resubmission issue

https://code.google.com/p/chromium/issues/detail?id=241888
https://code.google.com/p/chromium/issues/detail?id=244434

In Google Chrome for iOS the behavior of `window.history.back()` and `window.history.go(-1)` are
not equivalent. In case when page containing iframe submits a form to an iframe (via method POST)
the page URI is getting added to `window.history`, and the following call of `window.history.go(-1)`
causes the attempt to resubmit the form, and, consequently shows the "Confirm form resubmission" 
dialog, when `window.history.back()` behaves as expected.

Expected to be fixed in Chrome for iOS in v.28.

## Running Unit Tests

1. In `backbone-historytracker` directory run `node ./test/bin/server.js "" 1234`
2. In the browser navigate to `http://localhost:1234/test/index.html`


release-notes.md
# Release Notes

## Development

[Commits](https://github.com/walmartlabs/backbone-historytracker/compare/v1.0.4...master)

## v1.0.4 - October 10th, 2013

- [#7](https://github.com/walmartlabs/backbone-historytracker/pull/7) - User agent matches Non chrome browser ([@catesandrew](https://api.github.com/users/catesandrew))

[Commits](https://github.com/walmartlabs/backbone-historytracker/compare/v1.0.3...v1.0.4)

## v1.0.3 - August 30th, 2013

- [#6](https://github.com/walmartlabs/backbone-historytracker/pull/6) - Step out fix ([@Candid](https://api.github.com/users/Candid)

Compatibility notes:
- TODO : What might have broken?

[Commits](https://github.com/walmartlabs/backbone-historytracker/compare/v1.0.2...v1.0.3)


test/test.historytracker.js
/*global async */
$(document).ready(function() {

  var router = null;
  var lastRoute = null;
  var lastArgs = [];
  var steps = [];
  var startingIndex;

  function onRoute(router, route, args) {
    lastRoute = route;
    lastArgs = args;
  }

  function routeBind(callback) {
    var handler = function() {
      var args = arguments;
      Backbone.history.unbind('route', handler);

      // Let the route execute before we verify
      setTimeout(function() {
        callback.apply(undefined, args);
      }, 0);
    };
    Backbone.history.bind('route', handler);
  }

  function step(page, direction, offset, callback) {
    var index = steps.length;
    steps[index] = function(router, name, params) {
      var hist = Backbone.history;
      equals(hist.getFragment(), 'search/manhattan/p' + page);
      equals(params.direction, direction, 'route direction');
      equals(hist.getIndex(), startingIndex+offset, 'history index');
      equals(router.query, 'manhattan');
      equals(router.page, page);

      if (steps[index + 1]) {
        routeBind(steps[index + 1]);
      }
      setTimeout(callback, 0);
    };
  }

  function cleanup() {
    if (window.testPushState) {
      window.history.pushState({}, document.title, originalUrl);
    } else {
      window.location.hash = '';
    }
    start();
  }

  module("Backbone.Historytracker", {

    setup: function() {
      Backbone.history = null;
      router = new Router({testing: 101});
      Backbone.history.interval = 9;
      Backbone.history.start({pushState: window.testPushState, root: '/foo/', trackDirection: true});
      lastRoute = null;
      lastArgs = [];
      Backbone.history.on('route', onRoute);

      steps = [];
      startingIndex = Backbone.history.getIndex();
    },

    teardown: function() {
      Backbone.history.stop();
      Backbone.history.off('route', onRoute);
    }

  });

  var Router = Backbone.Router.extend({
    routes: {
      "search/:query":              "search",
      "search/:query/p:page":       "search"
    },

    search : function(query, page) {
      this.query = query;
      this.page = page;
    }
  });

  asyncTest("Router: index delta", 40, function() {
    step('20', 1, 1, function() {
      Backbone.history.navigate('search/manhattan/p30', {replace: true, trigger: true});
    });
    step('30', 0, 1, function() {
      Backbone.history.navigate('search/manhattan/p40', true);
    });
    step('40', 1, 2, function() {
      Backbone.history.back(true);
    });
    step('30', -1, 1, function() {
      Backbone.history.navigate('search/manhattan/p50', true);
    });
    step('50', 1, 2, function() {
      Backbone.history.navigate('search/manhattan/p60', true);
    });
    step('60', 1, 3, function() {
      Backbone.history.go(-2, false);
      Backbone.history.navigate('search/manhattan/p70', true);
    });
    step('70', 1, 2, function() {
      Backbone.history.back(true);
    });
    step('30', -1, 1, cleanup);

    routeBind(steps[0]);
    setTimeout(function() {
      Backbone.history.navigate('search/manhattan/p20', true);
    }, 0);
  });

  asyncTest("Router: navigate with fake replace", 3, function() {
    Backbone.history._fakeReplace = true;
    Backbone.history.navigate('search/q1', true);
    setTimeout(function() {
      equals(router.query, 'q1');
      Backbone.history.navigate('search/q2', true);
      setTimeout(function() {
        equals(router.query, 'q2');
        Backbone.history.navigate('search/q1', {trigger: true, replace: true});
        setTimeout(function() {
          equals(router.query, 'q1');
          Backbone.history._fakeReplace = false;
          start();
        }, 500);
      }, 0);
    }, 0);
  });

  asyncTest("Router: go ignore", 21, function() {
    var hist = Backbone.history;

    // Setup
    step('20', 1, 1, function() {
      hist.navigate('search/manhattan/p30', true);
    });
    step('30', 1, 2, function() {
      hist.navigate('search/manhattan/p40', true);
    });
    step('40', 1, 3, function() {
      hist.back(function(fragment) {
        equals(fragment, 'search/manhattan/p30', 'route returned value');

        setTimeout(function() {
          hist.back(function() { return true; });
        }, 10);
        return false;
      });
    });
    step('20', -1, 1, cleanup);

    routeBind(steps[0]);
    setTimeout(function() {
      hist.navigate('search/manhattan/p20', true);
    }, 0);
  });

  function execStepOut(options, test) {
    var hist = Backbone.history,
        iframe = $('<iframe id="foo-frame" src="data:text/html;base64, PGRpdj4xPC9kaXY+">'),
        form = $('<form action="" method="POST" target="foo-frame"></form>');

    var steps = [
      function() {
        hist.navigate('search/manhattan/p20', true);
      },
      function() {
        hist.navigate('search/manhattan/p30', true);
      },
      function() {
        hist.navigate('search/manhattan/p40', true);
        $('#qunit-fixture').append(iframe);
      }
    ];

    if (options.history) {
      steps.push(
        function() {
          iframe[0].setAttribute('src', 'data:text/html;base64, PGRpdj4yPC9kaXY+');
        },
        function() {
          iframe[0].setAttribute('src', 'data:text/html;base64, PGRpdj4zPC9kaXY+');
        },
        function() {
          iframe[0].setAttribute('src', 'data:text/html;base64, PGRpdj4yPC9kaXY+');
        });
    }

    if (options.emulateFormSubmit) {
      steps.push(function() {
        $('#qunit-fixture').append(form);
      });
      steps.push(function() {
        form.submit();
      });
    }

    steps.push(function() {
      equals(hist.getFragment(), 'search/manhattan/p40');
    });
    steps.push(test);

    steps = _.map(steps, function(step) {
      return function(callback) {
        setTimeout(function() {
          step();
          callback();
        }, 100);
      };
    });

    async.series(steps);
  }

  asyncTest("Router: stepOut", 2, function() {
    var hist = Backbone.history;

    execStepOut({history: true}, function() {
      hist.stepOut({
        view: window,
        callback: function() {
          equals(hist.getFragment(), 'search/manhattan/p30');
          start();
        }
      });
    });
  });

  asyncTest("Router: navigate after stepOut", function() {
    var hist = Backbone.history;

    execStepOut({history: true}, function() {
      hist.stepOut({
        trigger: true,
        view: window,
        callback: function() {
          equals(hist.getFragment(), 'search/manhattan/p30');

          setTimeout(function() {
            hist.navigate('search/manhattan/p40', {trigger: true});
            setTimeout(function() {
              equals(hist.getFragment(), 'search/manhattan/p40');
              start();
            }, 100);
          }, 100);
        }
      });
    });
  });

  asyncTest("Router: stepOut - trigger", 3, function() {
    var hist = Backbone.history;

    console.log('why you fail');
    execStepOut({history: true}, function() {
      hist.stepOut({
        view: window,
        trigger: 'search/manhattan/p20',
        routeLimit: 2,
        callback: function(fragment, trigger) {
          equals(fragment, 'search/manhattan/p20');
          equals(trigger, true);
          start();
        }
      });
    });
  });
  asyncTest("Router: stepOut - trigger limit", 3, function() {
    var hist = Backbone.history;

    console.log('but you work');
    execStepOut({history: true}, function() {
      hist.stepOut({
        view: window,
        trigger: 'search/manhattan/p10',
        routeLimit: 2,
        callback: function(fragment, trigger) {
          equals(fragment, 'search/manhattan/p10');
          equals(trigger, false);
          start();
        }
      });
    });
  });

  asyncTest("Router: stepOut - stepLimit", 2, function() {
    // Android 2.x does not insert iframe history into the global history which is helpful
    // for our desired behavior but means that we can't run tests like this.
    var android = navigator.userAgent.match(/Android\s+([\d.]+)/);
    if (android && parseFloat(android[1]) < 4) {
      expect(0);
      return start();
    }

    var hist = Backbone.history;

    execStepOut({history: true}, function() {
      hist.stepOut({
        view: window,
        stepLimit: 2,
        callback: function() {
          equals(hist.getFragment(), 'search/manhattan/p40');
          start();
        }
      });
    });
  });

  asyncTest("Router: stepOut - no iframe", 2, function() {
    var hist = Backbone.history;

    execStepOut({history: false}, function() {
      hist.stepOut({
        view: window,
        callback: function() {
          equals(hist.getFragment(), 'search/manhattan/p30');
          start();
        }
      });
    });
  });

  asyncTest("Router: stepOut - prevent form resubmission", function() {
    var hist = Backbone.history;

    execStepOut({history: false, emulateFormSubmit: true}, function() {
      hist.stepOut({
        view: window,
        callback: function() {
          equals(hist.getFragment(), 'search/manhattan/p30');
          start();
        }
      });
    });
  });
});


backbone.historytracker.js
(function() {
  // cached super methods
  var _route = Backbone.History.prototype.route;
  var _getFragment = Backbone.History.prototype.getFragment;
  var _start = Backbone.History.prototype.start;
  var _checkUrl = Backbone.History.prototype.checkUrl;
  var _navigate = Backbone.History.prototype.navigate;
  var _extractParameters = Backbone.Router.prototype._extractParameters;

  // If we are in hash mode figure out if we are on a browser that is hit by 63777 and 85881
  //     https://bugs.webkit.org/show_bug.cgi?id=63777
  var _useReplaceState = /WebKit\/([\d.]+)/.exec(navigator.userAgent) && window.history.replaceState,

      // Some variants of android will not replace the location history properly, causing issues
      // on back navigation. Hack around the replace API making it two step for these cases.
      //
      // This is likely tied to a specific Webkit version but sticking to the environments that we
      // know this occurs in rather than trying to track down the webkit version that this might
      // be impacting and possibly getting it wrong.
      //    https://bugs.webkit.org/show_bug.cgi?id=85881
      _fakeReplace = /Android\s+([\d.]+)((?!Chrome).)*$/.exec(navigator.userAgent) && (parseFloat(RegExp.$1) < 4.2);

  // pattern to recognize state index in hash
  var hashStrip = /^(?:#|%23)*\d*(?:#|%23)*/;
  // Cached regex for index extraction from the hash
  var indexMatch = /^(?:#|%23)*(\d+)(?:#|%23)/;

  _.extend(Backbone.Router.prototype, {
    // the direction index will be exposed on the parameters as 'direction'
    _extractParameters: function() {
      var params = _extractParameters.apply(this, arguments);
      params = params || [];
      var history = Backbone.history;
      if (history._trackDirection) {
        var oldIndex = history._directionIndex;
        history._directionIndex = history.loadIndex();
        params.direction = history._directionIndex - oldIndex;
      }
      return params;
    }
  });

  _.extend(Backbone.History.prototype, {
    // Get the location of the current route within the backbone history.
    // This should be considered a hint
    // Returns -1 if history is unknown or disabled
    getIndex : function() {
      return this._directionIndex;
    },

    getFragment : function(/* fragment, forcePushState */) {
      var rtn = _getFragment.apply(this, arguments);
      return rtn && rtn.replace(hashStrip, '');
    },

    start: function(/* options */) {
      var rtn = _start.apply(this, arguments);

      this._fakeReplace = _fakeReplace && !this.options.noReplaceHack;

      // Direction tracking setup
      this._trackDirection  = !!this.options.trackDirection;
      if (this._trackDirection) {
        var loadedIndex = this.loadIndex();
        this._directionIndex  = loadedIndex || window.history.length;
        this._state = {index: this._directionIndex};

        // If we are tracking direction ensure that we have a direction field to play with
        if (!loadedIndex) {
          var loc = window.location;
          if (!this._hasPushState) {
            loc.replace(loc.pathname + (loc.search || '') + '#' + this._directionIndex + '#' + this.fragment);
          } else {
            window.history.replaceState({index: this._directionIndex}, document.title, loc);
          }
        }
      }
      return rtn;
    },

    checkUrl : function(e) {
      var fromIframe = this.getFragment() == this.fragment && this.iframe;
      var loadedIndex = this.loadIndex(fromIframe && this.iframe.location.hash);
      var navigateOptions = {
        trigger: false,
        replace: !loadedIndex,
        forceIndex: loadedIndex || this._directionIndex + 1
      };
      _checkUrl.call(this, e, navigateOptions);
    },

    // The options object can contain `trigger: true` if you wish to have the
    // route callback be fired (not usually desirable), or `replace: true`, if
    // you wish to modify the current URL without adding an entry to the history.
    navigate: function(fragment, options) {
      if (this._ignoreChange) {
        this._pendingNavigate = _.bind(this.navigate, this, fragment, options);
        return;
      } else if (options && options.replace && this._fakeReplace) {
        // Under older Android environment we have to back out and then trigger a route in two
        // operations rather than an automic replace operation. Without this flow does not operate
        // correctly:
        //  1. Trigger #foo
        //  2. Replace #bar
        //  3. Trigger #baz
        //
        // This will trigger all of the proper routes but the history for this case will actually
        // be #foo, #baz rather than the desired #bar, #baz, causing unexpected navigation.
        options = _.omit(options, 'replace');
        options._replace = true;
        this._pendingNavigate = _.bind(this.navigate, this, fragment, options);
        this.back();
        return;
      }

      if (!options || options === true) {
        options = {trigger: options};
      }
      var newIndex;
      if (this._trackDirection) {
        newIndex = options.forceIndex || (this._directionIndex + (options.replace ? 0 : 1));

        if (options._replace) {
          // The step back mucked with our direction tracker. Update.
          this._directionIndex++;
        }
      }
      if (this._hasPushState) {
        options.state = {index: newIndex};
      } else {
        if (this._trackDirection) {
          fragment = newIndex + '#' + fragment;
        }
      }
      _navigate.call(this, fragment, options);
    },

    _updateHash: function(location, frag, replace) {
      var base = location.toString().replace(/(javascript:|#).*$/, '') + '#';
      if (replace) {
        if (_useReplaceState) {
          window.history.replaceState({}, document.title, base + frag);
        } else {
          location.replace(base + frag);
        }
      } else {
        location.hash = frag;
      }
    },

    // Pulls the direction index out of the state or hash
    loadIndex : function(fragmentOverride) {
      if (!this._trackDirection) {
        return;
      }
      if (!fragmentOverride && this._hasPushState) {
        return (this._state && this._state.index) || 0;
      } else {
        var match = indexMatch.exec(fragmentOverride || window.location.hash);
        return (match && parseInt(match[1], 10)) || 0;
      }
    },

    route: function (route, callback) {
      return _route.call(this, route, _.bind(function(fragment) {
        var ignore = this._ignoreChange;
        if (_.isFunction(ignore)) {
          ignore = !this._ignoreChange(fragment, route);
        }

        this._ignoreChange = false;

        if (ignore) {
          this._directionIndex = Backbone.history.loadIndex();
          var pendingNavigate = this._pendingNavigate;
          if (pendingNavigate) {
              var self = this;
            _.defer(function() {
              // back() was called before pendingNavigate and it could have updated
              // this.fragment to the same as pending navigate fragment. This will make
              // navigate a NOP. Hence resetting the fragment to make sure navigate will
              // do its job.
              self.fragment = undefined;
              pendingNavigate();
            });
          }
          this._pendingNavigate = undefined;
        } else {
          callback && callback.apply(this, arguments);
        }
      }, this));
    },

    back : function(triggerRoute) {
      this.go(-1, triggerRoute);
    },

    forward : function(triggerRoute) {
      this.go(1, triggerRoute);
    },

    go : function(count, triggerRoute) {
      this._ignoreChange = _.isFunction(triggerRoute) ? triggerRoute : !triggerRoute;

      // Explicitly use `back()` and `forward()` methods since `go(1)` and `go(-1)` may behave
      // unexpectedly in different browsers, although their behavior should be equivalent; see
      // the Chromium bugs https://code.google.com/p/chromium/issues/detail?id=241888 and
      // https://code.google.com/p/chromium/issues/detail?id=244434
      switch (count) {
        case -1:
          window.history.back();
          break;
        case 1:
          window.history.forward();
          break;
        default:
          window.history.go(count);
      }
    },

    /**
     * Hack to work around iframes that make have screwed with our history state.
     * Really one should choose not to use iframes but sometimes you must, even in 2013....
     */
    stepOut: function(options) {
      options = options || {};

      var stepLimit = options.stepLimit || 10,
          routeLimit,
          desiredRoute,
          iter = 0,
          timeout;

      if (options.view) {
        options.view.$('iframe').remove();
      }

      if (_.isString(options.trigger)) {
        // If a string is passed then we will ensure that they land on the desired route one way
        // or the other
        desiredRoute = options.trigger;
        routeLimit = options.routeLimit || 1;

        options.trigger = function(fragment) {
          routeLimit--;

          if (fragment === desiredRoute) {
            return true;
          } else if (routeLimit <= 0) {
            Backbone.history.navigate(desiredRoute, {replace: true, trigger: true});
            _.defer(function() {
              options.callback && options.callback(desiredRoute, false);
            });
          } else {
            _.defer(step);
          }
        };
      }

      // General flow here is to try to do a back navigation and wait for a backbone event to trigger
      // If one does not within a given timeout then repeat.
      function step() {
        iter++;

        // Timeout before as some envs actually have a sync callback for back. (Android 2.x notably)
        timeout = setTimeout(function() {
          if (iter > stepLimit) {
            options.callback && options.callback(false);
          } else {
            step();
          }
        }, 300);

        Backbone.history.back(function(fragment) {
          clearTimeout(timeout);

          var trigger = _.isFunction(options.trigger) ? options.trigger.apply(this, arguments) : options.trigger;
          if (trigger || !desiredRoute) {
            // Defer a callback since it can potentially call history's methods that the current execution would
            // interfere with (history methods rely on history state like history._ingoreChange)
            options.callback && _.defer(function() { options.callback(fragment, !!trigger); });
          }
          return trigger;
        });
      }

      step();
    }
  });

}());
